src/dune/action_exec.ml

```diff
@@ -181,6 +183,16 @@ let rec exec t ~ectx ~eenv =
   | Run (Ok prog, args) ->
     let+ () = exec_run ~ectx ~eenv prog args in
     Done
+  | With_exit_codes (pred, t) ->
+    let eenv =
+      { eenv with
+        exit_codes =
+          (fun i ->
+            Predicate_lang.Ast.exec pred
+              ~standard:(Predicate_lang.Ast.Element 0) (Int.equal i))
+      }
+    in
+    exec t ~ectx ~eenv
   | Dynamic_run (Error e, _) -> Action.Prog.Not_found.raise e
   | Dynamic_run (Ok prog, args) ->
     exec_run_dynamic_client ~ectx ~eenv prog args
```

Comment:
We should think a bit about what `exec_run_dynamic_client` should do.
It might be surprising if "With_exit_codes" configuration tells dune to ignore the error, but dune still complains with no extra explanation.
Maybe we could just explicitly disallow the combination of `Dynamic_run` and `With_exit_codes`. (or have them interact, but it's not obvious what's the right interaction)

Reply:
Good point. I added a dynamic check that fails if `(dynamic-run ...)` is used within the scope of `(with-exit-codes ...)`.